### PR TITLE
Add a safeguard to map command

### PR DESCRIPTION
**Pull request by catornot**
adds safeguard to the map command that prevents it from executing if the map is invalid.

I wanted to implement this originally as a plugin but I think it might be quite useful for everyone.

image

### CODE REVIEW

File: NorthstarDLL/util/printmaps.cpp
Comment (Alystrasz) [inline NorthstarDLL/util/printmaps.cpp:34]: I assume you meant `OriginalMapCommand` here :)

File: NorthstarDLL/util/printmaps.cpp
Comment (Alystrasz) [inline NorthstarDLL/util/printmaps.cpp:191]: Why is this required?

File: NorthstarDLL/util/printmaps.cpp
Comment (Alystrasz) [inline NorthstarDLL/util/printmaps.cpp:194]: I assume a map is deemed invalid if its name isn't contained in the loaded maps names, am I right?

**Review by Alystrasz — CHANGES_REQUESTED** [review]

File: NorthstarDLL/util/printmaps.cpp
Comment (ASpoonPlaysGames) [inline NorthstarDLL/util/printmaps.cpp:194]: `f` is rly not a good variable name, i assume it stands for `filter` or something?

File: NorthstarDLL/util/printmaps.cpp
Comment (F1F7Y) [reply NorthstarDLL/util/printmaps.cpp:191]: Refreshes the list of maps we can load. Imo this should be done in ModManager though that would be a separate PR

File: NorthstarDLL/util/printmaps.cpp
Comment (F1F7Y) [reply NorthstarDLL/util/printmaps.cpp:194]: Loading an invalid map ( can't be loaded for some reason ) causes an engine error

**Review by ASpoonPlaysGames — CHANGES_REQUESTED** [review]
Would be nice to get this merged, just a couple of questions/suggestions

File: NorthstarDLL/util/printmaps.cpp
Comment (Alystrasz) [reply NorthstarDLL/util/printmaps.cpp:197]: My vote goes to `warn`, since it's a user skill issue

File: NorthstarDLL/util/printmaps.cpp
Comment (H0L0theBard) [reply NorthstarDLL/util/printmaps.cpp:189]: If there isn't a map name, we should use the message which source uses for this which is `No map specified`.

File: NorthstarDLL/util/printmaps.cpp
Comment (H0L0theBard) [reply NorthstarDLL/util/printmaps.cpp:197]: a `warn` which is like `Map {name} does not exist` unless we want to log the same as regular source which is `map load failed: {name} not found or invalid`

File: NorthstarDLL/util/printmaps.cpp
Comment (catornot) [reply NorthstarDLL/util/printmaps.cpp:189]: `args.Arg(0)` is `map` so `> 1` is needed

File: NorthstarDLL/util/printmaps.cpp
Comment (catornot) [reply NorthstarDLL/util/printmaps.cpp:194]: I'll just put everything into one line

File: NorthstarDLL/util/printmaps.cpp
Comment (catornot) [reply NorthstarDLL/util/printmaps.cpp]: > If there isn't a map name, we should use the message which source uses for this which is `No map specified`.

the map concommand doesn't crash the engine when no arguments are provided.

Comment (F1F7Y) [reply NorthstarDLL/util/printmaps.cpp]: This is about feedback to the user, not preventing an engine crash

Comment (catornot) [reply NorthstarDLL/util/printmaps.cpp]: ok done

**Review by Alystrasz — APPROVED** [review]
Game does not crash anymore when passed an invalid map name, and prints a nice warning message instead.

**Review by F1F7Y — CHANGES_REQUESTED** [review]
I don't like how you replace the con command callback and then call the original. The original is just a few lines of code so I'd prefer we hook it and reconstruct it.

I can do this if you don't have the time to do so.

**Comment by catornot** [comment]
> I don't like how you replace the con command callback and then call the original. The original is just a few lines of code so I'd prefer we hook it and reconstruct it.

I can't find the function for it in ghidra 😞

**Comment by F1F7Y** [comment]
> I can't find the function for it in ghidra 😞

`engine.dll + 0x15B340`

image

**Comment by catornot** [comment]
ghidra decided to combine `Host_map_f` and `FUN_0015b220` lol

File: NorthstarDLL/util/printmaps.cpp
Comment (catornot) [reply NorthstarDLL/util/printmaps.cpp:212]: idk

**Review by Alystrasz — APPROVED** [review]
Still works as expected, game does not crash when command is given an incorrect map name.

File: NorthstarDLL/util/printmaps.cpp
Comment (Jan200101) [reply NorthstarDLL/util/printmaps.cpp:212]: `nullptr_t` is derived from `nullptr` so this is essentially a roundabout way of doing `nullptr`

**Review by ASpoonPlaysGames — APPROVED** [review]
Changes that I requested were addressed/are now outdated. Code looks good to me

**Review by Jan200101 — APPROVED** [review]
code looks good

extra check if the map actually exists may be overkill, but the vanilla map command is fragile so its probably best to guard it from inputs it may not be capable of handling

**Comment by ASpoonPlaysGames** [comment]
> extra check if the map actually exists may be overkill, but the vanilla map command is fragile so its probably best to guard it from inputs it may not be capable of handling

Game crashes if you try and do a map command with an invalid map

**Comment by Jan200101** [comment]
yeah my tired brain thought about the wrong thing.
I ment to say no map being specified at all and just running the original command if the inputs are good.
But the command is fragile so its best to gut it completely

**Review by GeckoEidechse — APPROVED** [review]
I tested the following inputs for the PR
```
map           <-- shows appropriate warning as expected
map mp_glich  <-- shows appropriate warning as expected
map mp_glitch <-- loads map as expected
```

without this PR what happens

```
map           <-- nothing happens
map mp_glich  <-- crashes game with error message about missing map
map mp_glitch <-- loads map as expected
```

Didn't do any code review as others have already covered that part anyway an I don't feel qualified enough apart from spotting basic errors.

**Comment by GeckoEidechse** [comment]
@F1F7Y I dismissed your review that requests changes as the requested changes have since been implemented as far as I can tell and the lack of a re-review was blocking the PR ^^

**Comment by snake-biscuits** [comment]
Has this been tested on maps which are in an installed mod but that mod is disabled?
Because those maps are listed when you type “map mp_” in console

Looking at the code changes it seems not…
Might test for myself & bug report / try a fix later

**Comment by GeckoEidechse** [comment]
This breaks docker image as per https://github.com/pg9182/northstar-dedicated/issues/70

**Pull request merged**
